### PR TITLE
feat(api-pagination): Integrating pagination on web3api

### PR DESCRIPTION
**Pull request by sogunshola**
---
name: 'Web3api pagination'
about: This pull request implements pagination when using web3api seamlessly by calling `await response.next()`
---

### Checklist

- [X] I am not disclosing a [vulnerability](https://github.com/MoralisWeb3/Moralis-JS-SDK/blob/main/SECURITY.md).
- [X] My code is conform the [code style](https://github.com/MoralisWeb3/Moralis-JS-SDK/blob/main/CODE_STYLE.md)
- [X] I have made corresponding changes to the documentation
- [X] I have updated Typescript definitions when needed

### Issue Description

Related issue: #`FILL_THIS_OUT`

### CODE REVIEW

File: scripts/genWeb3API.js [hunk 1]
```diff
@@ -137,7 +137,22 @@ static async fetchFromApi(endpoint, params) {
         'x-api-key': this.apiKey,
       },
     });
-    return response.data;
+    const result = response.data;
+    if(result.page_size) {
+      if(result.cursor) {
+        if(params.cursor === result.cursor) {
```
Comment (Capplequoppe) [inline scripts/genWeb3API.js:143]: Should we also have `&& params.cursor !== undefined`?

File: scripts/genWeb3API.js [hunk 2]
```diff
@@ -137,7 +137,22 @@ static async fetchFromApi(endpoint, params) {
         'x-api-key': this.apiKey,
       },
     });
-    return response.data;
+    const result = response.data;
```
Comment (Capplequoppe) [inline scripts/genWeb3API.js:140]: I think we can make this more DRY.
What about a helper function like this somewhere; 
```
function getNextOptions(result, options) {
  let nextOptions = options;
  const { result } = response.data;
  if (result.page_size) return options
  if (result.cursor) {
    if (nextOptions.cursor === result.cursor) return [];

    nextOptions.cursor = result.cursor;
  } else {
    if (result.total > (result.page_size * result.page) + 1) {
      nextOptions.offset = (result.page + 1) * (nextOptions.limit || 500);
    }
  }

  return options;
}
```

Then we can just assign the next function like this: 

```
const nextOptions = getNextOptions(result, options)
result.next = () => this.fetchFromServer(name, nextOptions);
```

or

```
const nextOptions = getNextOptions(result, options)
result.next = () => this.fetchFromApi(endpoint, nextOptions);
```

**Review by Capplequoppe — DISMISSED** [review]

File: scripts/genWeb3API.js [hunk 3]
```diff
@@ -79,6 +79,21 @@ static getParameterizedUrl(url, params) {
...
 }
 
+static getNextOptions(result, options) {
+  const nextOptions = options;
+  if (!result.page_size) return options
+  if (result.cursor && result.cursor !== undefined) {
+    if (nextOptions.cursor === result.cursor) return [];
+    nextOptions.cursor = result.cursor;
+  } else {
+    if (result.total > (result.page_size * result.page) + 1) {
```
Comment (ErnoW) [inline scripts/genWeb3API.js:89]: Shouldn't this be: `result.total > (result.page_size * (result.page + 1))`?

Comment (ErnoW) [reply scripts/genWeb3API.js:89]: We have no check if `result.total` or `result.page` is defined. This is probably the case when result.page_size is defined. But lets add the check to be safe

File: scripts/genWeb3API.js [hunk 4]
```diff
@@ -79,6 +79,21 @@ static getParameterizedUrl(url, params) {
   return parameterizedUrl;
 }
 
+static getNextOptions(result, options) {
+  const nextOptions = options;
+  if (!result.page_size) return options
+  if (result.cursor && result.cursor !== undefined) {
```
Comment (ErnoW) [inline scripts/genWeb3API.js:85]: No need to check for undefined if you check for `result.cursor`

File: scripts/genWeb3API.js [hunk 5]
```diff
@@ -137,7 +152,10 @@ static async fetchFromApi(endpoint, params) {
         'x-api-key': this.apiKey,
       },
     });
-    return response.data;
+    const result = response.data;
+    const nextOptions = this.getNextOptions(result, params)
+    result.next = () => this.fetchFromApi(endpoint, nextOptions);
```
Comment (ErnoW) [inline scripts/genWeb3API.js:157]: We don't always want to add a next function. We only want this for paginated requests (so when we use offset or cursor). Please add a conditional check for this.

**Review by ErnoW — CHANGES_REQUESTED** [review]

File: scripts/genWeb3API.js (hunk 4 above)
Comment (sogunshola) [reply scripts/genWeb3API.js:85]: I agree, but it this was a suggestion from Niclas.

**Comment by ErnoW** [comment]
Seems to be a use-case where the next() function is added, when it should not. See screenshot:
Screenshot 2022-04-25 at 13 55 29

Also
- Check the linting error
- add the next() function as optional Typescript type

**Comment by ErnoW** [comment]
Seems to be a use-case where the next() function is added, when it should not. See screenshot:
Screenshot 2022-04-25 at 13 55 29

Also
- Check the linting error
- add the next() function as optional Typescript type

**Review by ErnoW — APPROVED** [review]

**Pull request merged**